types/react-navigation/index.d.ts

```diff
@@ -303,18 +298,16 @@ export type NavigationAction =
   | NavigationStackAction
   | NavigationTabAction;
 
-export type NavigationRouteConfig<T> = T & {
+export type NavigationRouteConfig = NavigationComponent | ({
   navigationOptions?: NavigationScreenConfig<any>,
   path?: string,
-};
+} & NavigationScreenRouteConfig);
 
-export type NavigationScreenRouteConfig =
-  {
-      screen: NavigationComponent,
-    }
-  | {
-      getScreen: () => NavigationComponent,
-    };
+export type NavigationScreenRouteConfig = NavigationComponent | {
+    screen: NavigationComponent,
+} | {
+    getScreen: () => NavigationComponent,
```

Comment:
@timwangdev Here's where it was added. However getScreen isn't in the source code *directly*. Can check the search here https://github.com/react-navigation/react-navigation/search?q=getScreen&type=Code&utf8=✓. Looks like it's used in tests

Reply:
This change is add `NavigationCompoent` to the Union Type. It's still `screen` or `getScreen`. The test [here](https://github.com/react-navigation/react-navigation/blob/4569ad49f9224d7a980dafe3c3798fa2384aa788/src/routers/__tests__/validateRouteConfigMap-test.js#L40) says it should fail if `screen` and `getScreen` both defined.

Reply:
The test does that sure, but that's on the js side. I actually can't seem to get any combination of `screen` or `getScreen` to work.

I think the `NavigationComponent` is too strong of a typing as well. It doesn't allow for simple functional component since it extends from `React.ComponentClass`. Think we need to revert this

Reply:
Trying to force the type for a functional component fails as well: 
```tsx
const Lecturers: NavigationScreenComponent<{}, {}> = () => <Text>Lecturers</Text>
```

```
Type '() => Element' is not assignable to type 'NavigationScreenComponent<{}, {}>'.
  Type '() => Element' provides no match for the signature 'new (props: {}, context?: any): Component<{}, ComponentState>'.
```

Reply:
This is how it fails for the `getScreen`:

```tsx
export default StackNavigator(
  {
    lecturers: {
      screen: Lecturers,
    },
  }
)
```

```
Argument of type '{ lecturers: { screen: () => Element; }; }' is not assignable to parameter of type 'NavigationRouteConfigMap'.
  Property 'lecturers' is incompatible with index signature.
    Type '{ screen: () => Element; }' is not assignable to type 'NavigationRouteConfig'.
      Type '{ screen: () => Element; }' is not assignable to type '{ navigationOptions?: any; path?: string | undefined; } & { getScreen: () => NavigationComponent; }'.
        Type '{ screen: () => Element; }' is not assignable to type '{ getScreen: () => NavigationComponent; }'.
          Property 'getScreen' is missing in type '{ screen: () => Element; }'.
```

Reply:
TypeScript 2.7, using `strict: true`

Reply:
Just back from holiday. I will open a new PR for those issues later this week.

Reply:
@iRoachie Please check PR at #24471